Mathlib/Geometry/Manifold/Riemannian/PathELength.lean

```diff
@@ -0,0 +1,365 @@
+retaining the `C^1` property. We deduce that `riemannianEDist` satisfied the triangle inequality,
+in `riemannianEDist_triangle`.
+-/
+
+open Set MeasureTheory
+open scoped Manifold ENNReal ContDiff Topology
+
+noncomputable section
+
+variable
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E]
+  {H : Type*} [TopologicalSpace H] {I : ModelWithCorners ℝ E H} {n : WithTop ℕ∞}
+  {M : Type*} [TopologicalSpace M] [ChartedSpace H M]
+
+namespace Manifold
+
+variable [∀ (x : M), ENorm (TangentSpace I x)] {a b c a' b' : ℝ} {γ γ' : ℝ → M}
+
+variable (I) in
+/-- The length on `Icc a b` of a path into a manifold, where the path is defined on the whole real
+line.
+
+We use the whole real line to avoid subtype hell in API, but this is equivalent to
+considering functions on the manifold with boundary `Icc a b`, see
+`lintegral_norm_mfderiv_Icc_eq_pathELength_projIcc`.
+
+We use `mfderiv` instead of `mfderivWithin` in the definition as these coincide (apart from the two
+endpoints which have zero measure) and `mfderiv` is easier to manipulate. However, we give
+a lemma `pathELength_eq_integral_mfderivWithin_Icc` to rewrite with the `mfderivWithin` form. -/
+irreducible_def pathELength (γ : ℝ → M) (a b : ℝ) : ℝ≥0∞ :=
+  ∫⁻ t in Icc a b, ‖mfderiv 𝓘(ℝ) I γ t 1‖ₑ
+
+lemma pathELength_eq_lintegral_mfderiv_Icc :
+    pathELength I γ a b = ∫⁻ t in Icc a b, ‖mfderiv 𝓘(ℝ) I γ t 1‖ₑ := by simp [pathELength]
+
+lemma pathELength_eq_lintegral_mfderiv_Ioo :
+    pathELength I γ a b = ∫⁻ t in Ioo a b, ‖mfderiv 𝓘(ℝ) I γ t 1‖ₑ := by
+  rw [pathELength_eq_lintegral_mfderiv_Icc, restrict_Ioo_eq_restrict_Icc]
+
+lemma pathELength_eq_lintegral_mfderivWithin_Icc :
+    pathELength I γ a b = ∫⁻ t in Icc a b, ‖mfderivWithin 𝓘(ℝ) I γ (Icc a b) t 1‖ₑ := by
+  rw [pathELength_eq_lintegral_mfderiv_Icc, ← restrict_Ioo_eq_restrict_Icc]
+  apply setLIntegral_congr_fun measurableSet_Ioo (fun t ht ↦ ?_)
+  rw [mfderivWithin_of_mem_nhds]
+  exact Icc_mem_nhds ht.1 ht.2
+
+@[simp] lemma pathELength_self : pathELength I γ a a = 0 := by
+  simp [pathELength]
+
+lemma pathELength_congr_Ioo (h : EqOn γ γ' (Ioo a b)) :
+    pathELength I γ a b = pathELength I γ' a b := by
+  simp only [pathELength_eq_lintegral_mfderiv_Ioo]
+  apply setLIntegral_congr_fun measurableSet_Ioo (fun t ht ↦ ?_)
+  have A : γ t = γ' t := h ht
+  congr! 2
+  apply Filter.EventuallyEq.mfderiv_eq
+  filter_upwards [Ioo_mem_nhds ht.1 ht.2] with a ha using h ha
+
+lemma pathELength_congr (h : EqOn γ γ' (Icc a b)) : pathELength I γ a b = pathELength I γ' a b :=
+  pathELength_congr_Ioo (fun _ hx ↦ h ⟨hx.1.le, hx.2.le⟩)
+
+lemma pathELength_mono (h : a' ≤ a) (h' : b ≤ b') :
+    pathELength I γ a b ≤ pathELength I γ a' b' := by
+  simp only [pathELength_eq_lintegral_mfderiv_Icc]
+  exact lintegral_mono_set (Icc_subset_Icc h h')
+
+lemma pathELength_add (h : a ≤ b) (h' : b ≤ c) :
+    pathELength I γ a b + pathELength I γ b c = pathELength I γ a c := by
+  symm
```

Comment:
I'm not sure if this would simplify by working in the other direction instead (on this and the next line): in any case, optional.